source/time_loop_mgrit.h

```diff
@@ -0,0 +1,210 @@
+
+        void output(const vector_type &U,
+            const std::string &name,
+            Number t,
+            unsigned int cycle);
+
+        void print_parameters(std::ostream &stream);
+        void print_mpi_partition(std::ostream &stream);
+        void print_memory_statistics(std::ostream &stream);
+        void print_timers(std::ostream &stream);
+        void print_throughput(unsigned int cycle,
+            Number t,
+            std::ostream &stream,
+            bool final_time = false);
+
+        void print_info(const std::string &header);
+        void print_head(const std::string &header,
+            const std::string &secondary,
+            std::ostream &stream);
+
+        void print_cycle_statistics(unsigned int cycle,
+            Number t,
+            unsigned int output_cycle,
+            bool write_to_logfile = false,
+            bool final_time = false);
+        //@}
+
+      private:
+        /**
+         * @name Run time options
+         */
+        //@{
+
+        std::string base_name_;
+
+        Number t_initial_;
+        Number t_final_;
+        vector_type U_;
+
+        std::vector<Number> t_refinements_;
+
+        Number output_granularity_;
+
+        bool enable_checkpointing_;
+        bool enable_output_full_;
+        bool enable_output_levelsets_;
+        bool enable_compute_error_;
+        bool enable_compute_quantities_;
+        bool post_process_;
+
+        unsigned int output_checkpoint_multiplier_;
+        unsigned int output_full_multiplier_;
+        unsigned int output_levelsets_multiplier_;
+        unsigned int output_quantities_multiplier_;
+
+        std::vector<std::string> error_quantities_;
+        bool error_normalize_;
+
+        bool resume_;
+
+        Number terminal_update_interval_;
+        bool terminal_show_rank_throughput_;
+
+        //@}
+        /**
+         * @name Internal data:
+         */
+        //@{
+
+        const MPI_Comm &mpi_communicator_;
+
+        using map_type = typename std::map<std::string, dealii::Timer>;
+        map_type computing_timer_;
+
+        //todo: add pointers here.
+        std::shared_ptr<HyperbolicSystem> hyperbolic_system_;
+        std::shared_ptr<ParabolicSystem> parabolic_system_;
+        std::shared_ptr<Discretization<dim>> discretization_;
+        std::shared_ptr<OfflineData<dim, Number>> offline_data_;
+        std::shared_ptr<InitialValues<Description, dim, Number>> initial_values_;
+        std::shared_ptr<HyperbolicModule<Description, dim, Number>> hyperbolic_module_;
+        std::shared_ptr<ParabolicModule<Description, dim, Number>> parabolic_module_;
+        std::shared_ptr<TimeIntegrator<Description, dim, Number>> time_integrator_;
+        std::shared_ptr<Postprocessor<Description, dim, Number>> postprocessor_;
+        std::shared_ptr<VTUOutput<Description, dim, Number>> vtu_output_;
+        std::shared_ptr<Quantities<Description, dim, Number>> quantities_;
```

Comment:
You have all these objects already in `LevelStructures`. Does it make sense to instead store a reference or pointer to an object of `LevelStructures`?

Reply:
Make them `shared_ptr<const HyperbolicSystem etc>`